File: Javascript/Algorithms/94.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(node) {
  const stack = [], values = [];
  while (stack.length > 0 || node) {
    // keep going left until possible.
    if (node) {
      stack.push(node);
      node = node.left;
    } else {
      // pop left and go right
      node = stack.pop();
      values.push(node.val);
      node = node.right;
    }
  }
  return values;
};

File: Python/Algorithms/94.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def inorderTraversal(self, node):
        stack = []
        while stack or (node is not None):
            # keep going left until possible.
            if node is not None:
                stack.append(node)
                node = node.left
            else:
                # pop left and go right
                node = stack.pop()
                yield node.val
                node = node.right
